src/include/dawg/residue.h

```diff
@@ -58,343 +56,380 @@ struct alignment : public std::vector<details::aligned_sequence> {
+            return DAWG_ERROR("invalid sequence type");
+        }
+        cs_encode_ = &rmods[type_ * 80];
+        cs_ins_ = &sIns[(markins_ ? 1 : 0)];
+
+        gap_ = static_cast<unsigned int>(strchr(cs_decode_, '-') - cs_decode_);
+
+        do_op_append = (type_ == CODON) ? &residue_exchange::do_op_append_cod
+                                        : &residue_exchange::do_op_append_res;
+        do_op_appendi = (type_ == CODON) ? &residue_exchange::do_op_appendi_cod
+                                         : &residue_exchange::do_op_appendi_res;
+
+        return true;
+    };
+    inline bool is_same_type(unsigned int type, bool markins,
+                             bool keepempty) const {
+        return (type == type_ && markins == markins_ &&
+                keepempty == keepempty_);
+    }
+    inline bool is_same_model(unsigned int type, unsigned int code, bool rna,
+                              bool lowercase, bool markins,
+                              bool keepempty) const {
+        return (type == type_ && code == code_ && rna == rna_ &&
+                lowercase == lowercase_ && markins == markins_ &&
+                keepempty == keepempty_);
+    }
+    inline bool is_keep_empty() const { return keepempty_; }
+
+    inline residue::data_type gap_base() const { return gap_; }
+
+    inline residue::data_type encode(char ch) const {
+        char ret = (ch >= '0') ? (cs_encode_[ch - '0']) : -1;
+        return static_cast<residue::data_type>(ret);
+    }
+
+    ///////////////////////////////////////////////////////////
+    /// \brief encode
+    /// \param root_seq The root sequence to encode
+    ///////////////////////////////////////////////////////////
+    inline sequence encode(const std::string &root_seq) const {
+        static const auto getCodonNumber = [](const char a, const char b,
+                                              const char c) -> unsigned int {
+            return a | (b << 8) | (c << 16);
+        };
+        sequence residues;
+        if(type_ != CODON) {
+            for(size_t i = 0; i != root_seq.size(); ++i) {
+                auto base = encode(root_seq.at(i));
+                if(base == static_cast<decltype(base)>(-1)) {
+                    DAWG_ERROR("Invalid user sequence");
+                    return {};
+                }
+                residues.emplace_back(base, 0, 0);
+            }
+        } else {
+            if(root_seq.size() % 3 != 0) {
+                DAWG_ERROR(
+                    "Invalid user sequence, sequence does not fit codon.");
+                return {};
+            }
+            for(size_t i = 0; i + 2 < root_seq.size(); i += 3) {
+                residues.emplace_back(triplet_to_codon(getCodonNumber(
+                                          root_seq.at(i), root_seq.at(i + 1),
+                                          root_seq.at(i + 2))),
+                                      0, 0);
+                if(residues.back().data() > 63) {
```

Comment:
@reedacartwright the old comparison was `residues.back().data() == -1` but that compares and unsigned int to -1. I updated it to this because to my understanding `triplet_to_codon` should return a number [0,63]. I would like your confirmation for this change.